### PR TITLE
Added InvalidPageError, and raise MalformedPDFError when InvalidPageError is raised internally

### LINKED ISSUE #220: Malformed PDF raises `invalid page: 2 (ArgumentError)`
pdf-reader crashes with the attached malformed PDF file.

[20170708113958670209722_test_0109.pdf](https://github.com/yob/pdf-reader/files/1133081/20170708113958670209722_test_0109.pdf)

### Crash

```
[*] Processing 'crashes/20170708113958670209722_test_0109.pdf'
[+] Processing complete
[*] Parsing 'crashes/20170708113958670209722_test_0109.pdf'
[*] Version: 1.4
[*] Info: {:Creator=>"Writer", :Producer=>"OpenOffice.org 2.0", :CreationDate=>"D:20080409174640-04'00'"}
[*] Metadata: 
[*] Objects: <PDF::Reader::ObjectHash size: 84>
[*] Pages: 7
[*] Dumping PDF contents...
/usr/local/rvm/gems/ruby-2.3.0/gems/pdf-reader-2.0.0/lib/pdf/reader/page.rb:39:in `initialize': invalid page: 2 (ArgumentError)
	from /usr/local/rvm/gems/ruby-2.3.0/gems/pdf-reader-2.0.0/lib/pdf/reader.rb:177:in `new'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/pdf-reader-2.0.0/lib/pdf/reader.rb:177:in `block in pages'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/pdf-reader-2.0.0/lib/pdf/reader.rb:176:in `each'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/pdf-reader-2.0.0/lib/pdf/reader.rb:176:in `map'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/pdf-reader-2.0.0/lib/pdf/reader.rb:176:in `pages'
	from ./read-pdf.rb:78:in `parse'
	from ./read-pdf.rb:55:in `read'
	from ./read-pdf.rb:96:in `<main>'
```

### Reproduce

```
#!/usr/bin/env ruby
require 'pdf-reader'
require 'colorize'

$QUIET = false

#
# Print status message
#
# @param [String] msg message to print
#
def print_status(msg = '')
  puts '[*] '.blue + msg unless $QUIET
end

#
# Print progress messages
#
# @param [String] msg message to print
#
def print_good(msg = '')
  puts '[+] '.green + msg unless $QUIET
end

#
# Print error message
#
# @param [String] msg message to print
#
def print_error(msg = '')
  puts '[-] '.red + msg unless $QUIET
end

#
# Read a PDF file
#
# @param [File] file PDF file
#
def read doc
  print_status "Processing '#{doc}'"
  begin
    @fname = doc
    reader = PDF::Reader.new(@fname)
  rescue PDF::Reader::MalformedPDFError
    print_error "Could not parse PDF '#{doc}': PDF is malformed"
    exit 1
  rescue PDF::Reader::UnsupportedFeatureError 
    print_error "Could not parse PDF '#{doc}': PDF::Reader::UnsupportedFeatureError"
    exit 1
  end
  print_good 'Processing complete'

  print_status "Parsing '#{@fname}'"
  begin
    parse reader
  rescue PDF::Reader::UnsupportedFeatureError
    print_error "Could not parse PDF '#{doc}': PDF::Reader::UnsupportedFeatureError"
    exit 1
  rescue PDF::Reader::MalformedPDFError => e
    print_error "Could not parse PDF '#{doc}': PDF is malformed"
    exit 1
  end
  print_good 'Parsing complete'
end

#
# parse pdf
#
def parse(reader)
  print_status "Version: #{reader.pdf_version}"
  print_status "Info: #{reader.info}"
  print_status "Metadata: #{reader.metadata}"
  print_status "Objects: #{reader.objects}"
  print_status "Pages: #{reader.page_count}"

  print_status 'Dumping PDF contents...'
  contents = ''
  reader.pages.each do |page|
    contents << page.text.to_s
    contents << page.fonts.to_s
    contents << page.text.to_s
    contents << page.raw_content.to_s
  end
  #puts contents unless $QUIET
end

def usage
  print_status "./read-pdf.rb <FILE>"
  exit 1
end

doc = ARGV[0]
usage if doc.nil?

if File.exist? doc
  read doc
  exit 0
else
  print_error "Could not find #{doc}"
  exit 1
end
```

### PR DESCRIPTION
**Pull request by ndbroadbent**
* Raise InvalidPageError when an invalid page number is requested. 
* Raise MalformedPDFError when this happens inside the #pages iterator.

Fixes #220

### CODE REVIEW

File: spec/integrity.yml [hunk 1]
```diff
@@ -169,10 +169,13 @@ data/inline_image.pdf:
   :md5: b02bfbb6b0ee7c7d7df1b5f0c1f198c1
 data/inline_image_single_line_content_stream.pdf:
   :bytes: 130786
-  :md5: 096fb28baf29a716066768cb31182b73
+  :md5: '096fb28baf29a716066768cb31182b73'
```
Comment (yob) [inline spec/integrity.yml]: as in #256, it'd be good if the diffs avoided unrelated changes like this

File: spec/integrity_spec.rb [hunk 2]
```diff
@@ -8,7 +8,7 @@
 # line endings) then specs wil fail in confusing ways.
 #
 # This spec will fail if a new PDF is added to the suite but not
-# listed in spec/inhtegrity.yml. After adding the new PDF, be sure to
+# listed in spec/integrity.yml. After adding the new PDF, be sure to
```
Comment (yob) [inline spec/integrity_spec.rb]: thanks :)

File: lib/pdf/reader/error.rb [hunk 3]
```diff
@@ -53,6 +53,10 @@ def self.assert_equal(lvalue, rvalue)
   class MalformedPDFError < RuntimeError; end
 
   ################################################################################
+  # an exception that is raised when an invalid page number is used
+  class InvalidPageError < RuntimeError; end
```
Comment (yob) [inline lib/pdf/reader/error.rb]: This new error is generally raised when someone requests an invalid page number - are you open to changing this to inherit from `ArgumentError`?

File: spec/reader_spec.rb [hunk 4]
```diff
@@ -126,6 +131,11 @@
     it "returns a single page from no_text_spaces" do
       expect(PDF::Reader.new(no_text_spaces).page(1)).to be_a_kind_of(PDF::Reader::Page)
     end
+
+    it "raises InvalidPageError when an invalid page number is requested" do
+      reader = PDF::Reader.new(pdf_spec_file("cairo-basic"))
+      expect { reader.page(10) }.to raise_error(PDF::Reader::InvalidPageError)
+    end
```
Comment (yob) [inline spec/reader_spec.rb]: These specs look good, but can the first one be duplicated into `spec/integration_spec.rb`? As a file-parsing lib, I rely on that set of integration specs across the growing corpus of test files to avoid regressions.

File: spec/integrity.yml (hunk 1 above)
Comment (ndbroadbent) [reply spec/integrity.yml]: 👍

File: lib/pdf/reader/error.rb (hunk 3 above)
Comment (ndbroadbent) [reply lib/pdf/reader/error.rb]: Yes, good point

File: spec/reader_spec.rb (hunk 4 above)
Comment (ndbroadbent) [reply spec/reader_spec.rb]: 👍

**Comment by ndbroadbent** [comment]
Have pushed all those changes

**Comment by yob** [comment]
Thanks for this. I rebased it to resolve the merge conflict and it's been merged into master as 836d8b3.
